// @mr-hope/vuepress-plugin-comment/lib/client/TagInfo.js
import Vue from "vue";
import TagIcon from "./icons/TagIcon.vue";
import { capitalize } from "@mr-hope/vuepress-shared";
import { pageInfoI18n } from "./define";
export default Vue.extend({
    name: "TagInfo",
    components: { TagIcon },
    props: {
        tags: { type: Array, default: () => [] },
    },
    computed: {
        $tags() {
            if (this.tags.length !== 0)
                return this.tags;
            const { tags, tag = tags } = this.$frontmatter;
            if (typeof tag === "string")
                return [capitalize(tag)];
            if (Array.isArray(tag))
                return tag.map((item) => capitalize(item));
            return [];
        },
        clickable() {
            return this.$themeConfig.blog !== false;
        },
        hint() {
            return pageInfoI18n[this.$localePath || "/"].tag;
        },
    },
    methods: {
        navigate(tagName) {
            const path = `/tag/${tagName}/`;
            if (this.$route.path !== path)
                void this.$router.push(path);
        },
    },
});
