// @mr-hope/vuepress-plugin-comment/lib/node/index.js
"use strict";
const vuepress_shared_1 = require("@mr-hope/vuepress-shared");
const i18n_1 = require("./i18n");
module.exports = (options, { themeConfig }) => {
    const baseLang = options.baseLang || themeConfig.baseLang || "en-US";
    const baseLangPath = vuepress_shared_1.lang2Path(baseLang);
    const pageInfoI18nConfig = i18n_1.pageInfoI18n;
    const valineI18nConfig = i18n_1.valineI18n;
    pageInfoI18nConfig["/"] = pageInfoI18nConfig[baseLangPath];
    valineI18nConfig["/"] = valineI18nConfig[baseLangPath];
    const config = {
        name: "comment",
        define: () => ({
            COMMENT_OPTIONS: Object.keys(options).length > 0 ? options : themeConfig.comment || {},
            PAGE_INFO_I18N: pageInfoI18nConfig,
            VALINE_I18N: valineI18nConfig,
        }),
        plugins: [
            ["@mr-hope/git", themeConfig.git || true],
            ["@mr-hope/reading-time", { wordPerminute: options.wordPerminute }],
        ],
    };
    if (options.type === "vssue")
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        config.plugins.push(["@vssue/vuepress-plugin-vssue", options]);
    return config;
};
