// @mr-hope/vuepress-plugin-feed/lib/node/feed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Feed = void 0;
const atom_1 = require("./atom");
const json_1 = require("./json");
const rss_1 = require("./rss");
class Feed {
    constructor(options) {
        this.options = options;
        this.items = [];
        this.categories = new Set();
        this._contributorKeys = new Set();
        this.contributors = [];
        /**
         * Add a feed item
         */
        this.addItem = (item) => {
            this.items.push(item);
        };
        /**
         * Add a category
         */
        this.addCategory = (category) => {
            this.categories.add(category);
        };
        /**
         * Add a contributor
         */
        this.addContributor = (contributor) => {
            // use keys to avoid adding same contributor
            const key = contributor.email || contributor.name;
            if (key && !this._contributorKeys.has(key)) {
                this._contributorKeys.add(key);
                this.contributors.push(contributor);
            }
        };
        /**
         * Returns a Atom 1.0 feed
         */
        this.atom = () => atom_1.renderAtom(this);
        /**
         * Returns a RSS 2.0 feed
         */
        this.rss = () => rss_1.renderRSS(this);
        /**
         * Returns a JSON 1.1 feed
         */
        this.json = () => json_1.renderJSON(this);
    }
}
exports.Feed = Feed;
