// @mr-hope/vuepress-plugin-feed/lib/node/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFeedLinks = exports.getFeedChannelOption = exports.getOutput = exports.checkOptions = void 0;
const vuepress_shared_1 = require("@mr-hope/vuepress-shared");
const utils_1 = require("./utils");
const checkOptions = (options, context) => {
    const baseLang = options.baseLang || context.themeConfig.baseLang;
    const hostname = options.hostname || context.themeConfig.hostname;
    // make sure hostname do not end with `/`
    if (hostname)
        options.hostname = hostname.replace(/\/?$/u, "");
    else {
        utils_1.error("Option 'hostname' is required!");
        return false;
    }
    options.baseLang = baseLang || "en-US";
    return true;
};
exports.checkOptions = checkOptions;
const getOutput = (output) => {
    const defaultOption = {
        atom: {
            enable: true,
            path: "atom.xml",
        },
        json: {
            enable: true,
            path: "feed.json",
        },
        rss: {
            enable: true,
            path: "rss.xml",
        },
    };
    return vuepress_shared_1.deepAssign(defaultOption, output || {});
};
exports.getOutput = getOutput;
const getFeedChannelOption = (options, context) => {
    const { baseLang, hostname, icon, image } = options;
    const { base, themeConfig } = context;
    const { title, description } = context.getSiteData();
    const author = options.channel?.author?.name || themeConfig.author;
    const copyright = themeConfig.footer?.copyright || (author ? `Copyright by ${author}` : "");
    const defaultChannelOpion = {
        title,
        link: utils_1.resolveUrl(hostname, base),
        description,
        language: baseLang,
        copyright,
        pubDate: new Date(),
        lastUpdated: new Date(),
        ...(icon ? { icon } : {}),
        ...(image ? { image } : {}),
        ...(author
            ? {
                author: {
                    name: author,
                },
            }
            : {}),
    };
    return vuepress_shared_1.deepAssign(defaultChannelOpion, options.channel || {});
};
exports.getFeedChannelOption = getFeedChannelOption;
const getFeedLinks = (options, output, context) => {
    const { base } = context;
    const { hostname } = options;
    return {
        atom: utils_1.resolveUrl(hostname, base, output.atom.path),
        json: utils_1.resolveUrl(hostname, base, output.json.path),
        rss: utils_1.resolveUrl(hostname, base, output.rss.path),
    };
};
exports.getFeedLinks = getFeedLinks;
