// @mr-hope/vuepress-plugin-pwa/lib/node/index.js
"use strict";
const vuepress_shared_1 = require("@mr-hope/vuepress-shared");
const path_1 = require("path");
const i18n_1 = require("./i18n");
const injectHead_1 = require("./injectHead");
const genManifest_1 = require("./genManifest");
const genServiceWorker_1 = require("./genServiceWorker");
const pwaPlugin = (options, context) => {
    const { base, themeConfig } = context;
    const baseLang = options.baseLang || themeConfig.baseLang || "en-US";
    const baseLangPath = vuepress_shared_1.lang2Path(baseLang);
    const pwaI18nConfig = i18n_1.i18n;
    const pwaOption = Object.keys(options).length > 0 ? options : themeConfig.pwa || {};
    pwaI18nConfig["/"] = pwaI18nConfig[baseLangPath];
    const config = {
        name: "pwa",
        define: () => ({
            PWA_I18N: pwaI18nConfig,
            SW_BASE_URL: base || "/",
        }),
        globalUIComponents: [pwaOption.popupComponent || "SWUpdatePopup"],
        enhanceAppFiles: path_1.resolve(__dirname, "../client/enhanceAppFile.js"),
        beforeDevServer(app) {
            app.get(`${base || "/"}manifest.webmanifest`, (_req, res) => {
                genManifest_1.getManifest(pwaOption, context)
                    .then((manifest) => {
                    res.send(manifest);
                })
                    .catch(() => res.status(500).send("Unexpected manifest generate error"));
            });
        },
        ready() {
            context.siteConfig.head = injectHead_1.injectLinkstoHead(pwaOption, base, context.siteConfig.head);
        },
        async generated() {
            await genManifest_1.genManifest(pwaOption, context);
            await genServiceWorker_1.genServiceWorker(pwaOption, context);
        },
    };
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (pwaOption.showInstall !== false)
        config.globalUIComponents.push("PWAInstall");
    return config;
};
module.exports = pwaPlugin;
