// vuepress-plugin-photo-swipe/lib/node/index.js
"use strict";
const path_1 = require("path");
const vuepress_shared_1 = require("@mr-hope/vuepress-shared");
const i18n_1 = require("./i18n");
module.exports = (options, { themeConfig }) => {
    const option = Object.keys(options).length > 0
        ? options
        : themeConfig.photoSwipe || {};
    const baseLang = options.baseLang || themeConfig.baseLang || "en-US";
    const baseLangPath = vuepress_shared_1.lang2Path(baseLang);
    const photoSwipeI18nConfig = i18n_1.i18n;
    photoSwipeI18nConfig["/"] = photoSwipeI18nConfig[baseLangPath];
    return {
        name: "photo-swipe",
        define: () => ({
            IMAGE_CONTAINER: option.container || ".theme-default-content",
            IMAGE_SELECTOR: option.selector || ".theme-default-content :not(a) > img",
            PHOTOSWIPE_OPTIONS: option.options || {},
            PHOTOSWIPE_I18N: photoSwipeI18nConfig,
        }),
        enhanceAppFiles: path_1.resolve(__dirname, "../client/enhanceAppFile.js"),
        globalUIComponents: "PhotoSwipeUI",
    };
};
